examples/recording_service/leveldb-integration/cpp/LevelDbReader.cxx

```diff
@@ -0,0 +1,637 @@
+     * used as a key
+     */
+    const StructType& key_type = rti::topic::dynamic_type<UserDataKey>::get();
+    key_buffer_.resize(key_type.cdr_serialized_sample_max_size());
+
+    db_iterator_.reset(discovery_db_->NewIterator(leveldb::ReadOptions()));
+    for (db_iterator_->SeekToFirst(); db_iterator_->Valid(); ) {
+        UserDataKey current_key = slice_to_user_type<UserDataKey>(
+                db_iterator_->key(),
+                key_buffer_);
+        current_timestamp_ = current_key.reception_timestamp();
+        if (current_timestamp_ >= start_timestamp_
+                && current_timestamp_ <= end_timestamp_) {
+            break;
+        } else {
+            db_iterator_->Next();
+        }
+    }
+    /* The DB may be empty (no sample fulfills the conditions) */
+    if (!db_iterator_->Valid()) {
+        finished_ = true;
+    }
+
+    std::stringstream metadata_filename;
+    metadata_filename << working_dir << "/metadata.dat";
+    db_status = leveldb::DB::Open(
+            leveldb::Options(),
+            metadata_filename.str(),
+            &db);
+    if (!db_status.ok() || db == nullptr) {
+        std::stringstream log_stream;
+        log_stream << "Failed to open metadata DB file" << std::endl;
+        log_stream << "    File name = " << metadata_filename.str()
+                << std::endl;
+        log_stream << "    LevelDB error = " << db_status.ToString()
+                << std::endl;
+        throw std::runtime_error(log_stream.str());
+    }
+    metadata_db_.reset(db);
+
+    std::stringstream timestamp_stream;
+    std::string timestamp_str;
+
+    db_status = metadata_db_->Get(
+            leveldb::ReadOptions(),
+            start_time_key_name(),
+            &timestamp_str);
+    if (!db_status.ok()) {
+        std::stringstream log_stream;
+        log_stream << "Failed to read start time to metadata DB" << std::endl;
+        log_stream << "    LevelDB error = " << db_status.ToString()
+                << std::endl;
+        throw std::runtime_error(log_stream.str());
+    }
+    timestamp_stream.str(timestamp_str);
+    timestamp_stream >> service_start_time_;
+
+    db_status = metadata_db_->Get(
+            leveldb::ReadOptions(),
+            end_time_key_name(),
+            &timestamp_str);
+    if (!db_status.ok()) {
+        std::stringstream log_stream;
+        log_stream << "Failed to read start time to metadata DB" << std::endl;
+        log_stream << "    LevelDB error = " << db_status.ToString()
+                << std::endl;
+        throw std::runtime_error(log_stream.str());
+    }
+    timestamp_stream.str(timestamp_str);
+    timestamp_stream >> service_end_time_;
+}
+
+LevelDbStreamInfoReader::~LevelDbStreamInfoReader()
+{
+}
+
+void LevelDbStreamInfoReader::read(
+        std::vector<rti::routing::StreamInfo *>& sample_seq,
+        const rti::recording::storage::SelectorState& selector)
+{
+    if (finished_) {
```

Comment:
why is not Replay doing this independently of the plugin?

Reply:
I think I documented this in the issue that I created with the common behavioural patterns to implement as helpers.